Create route53 cname

chbatey opened the pull request:
Hi @justinsb this PR adds a CNAME for an ELB is you provide a hosted zone annotation. Where as we realise that DNS is a bigger story and likely part of ingress we will think there is value in adding a CNAME for an ELB. We've marked the annocation as beta as with the internal ELB. The format of the cname is <servicename>-<namespace>-<cluster-name>.<hosted-zone>

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/23576)

<!-- Reviewable:end -->

Commit:
Create route53 cname for ELBs

For the AWS provider, if a service has label:

    kubernetes.io/aws-lb-cname-zone: k8s.api.com

Then a cname will be generated for the elastic load balancer in the
hosted zone `k8s.api.com.`. The cname prefix is
`service-namespace-cluster`, e.g.

    kubedash-kube-system-mycluster.k8s.api.com

The benefit is that services get stable external DNS entries. The ELB
hostname is not reliable for external dependents, as it is autogenerated
by AWS and can change whenever the service is recreated.

Note: The aws client rejects concurrent cname creation with
'PriorRequestNotComplete'. This change uses a mutex to ensure only a
single cname is created at a time.

therc commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -308,3 +310,82 @@ func (s *AWSCloud) ensureLoadBalancerInstances(loadBalancerName string, lbInstan
+					ResourceRecordSet: &route53.ResourceRecordSet{
+						Name:            aws.String(cname),
+						Type:            aws.String("CNAME"),
+						TTL:             aws.Int64(600),
```
Why 600? Route53's default is 300.

ajohnstone commented:
Why CNAME and not an ALIAS?

chbatey commented:
@ajohnstone yes I think it makes sense to make it an ALIAS

@therc we can leave the TTL as the default

@justinsb are you happy with merging this if we make the changes suggested?

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -158,6 +164,12 @@ type ASG interface {
 }
 
+// This is a simple pass-through of the route 53 client interface, which allows for testing
+type R53 interface {
```
I worry `R53` is a little obscure.  Could we either name this Route53 or DNS?

chbatey replied:
Done

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -343,6 +357,14 @@ func (p *awsSDKProvider) Autoscaling(regionName string) (ASG, error) {
+		Region:      &regionName,
+		Credentials: p.creds,
+	}))
+	return r53Client, nil
```
Let's either include the call to `addHandlers` or have a note as to why not...

.. I realize the same applies to Metadata, but that's a separate issue ;-)

chbatey replied:
Done

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -2260,20 +2294,32 @@ func (s *AWSCloud) GetLoadBalancer(service *api.Service) (*api.LoadBalancerStatu
-		ingress.Hostname = orEmpty(lb.DNSName)
-		status.Ingress = []api.LoadBalancerIngress{ingress}
+		var dnsIngress api.LoadBalancerIngress
+		dnsIngress.Hostname = orEmpty(lb.DNSName)
```
`aws.StringValue` does the same as `orEmpty`.  I try to fix when I'm touching lines that orEmpty.  Up to you whether you think it's worth it here.

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -2260,20 +2294,32 @@ func (s *AWSCloud) GetLoadBalancer(service *api.Service) (*api.LoadBalancerStatu
 
-	return status
+	cname := s.getLoadBalancerCname(apiService)
+	glog.Info("CName for service %v is %v", apiService.Name, cname)
```
Info is pretty high priority for a purely informative message.  I suggest `glog.V(4).Infof`

Also note that it must be Infof (not Info) to format (I'm always making this mistake!)

Also, I like to use %q, which is like %s but wraps the string in quotes - here because cname might be the empty string it would be much clearer to use %q.  Though I wonder if we should actually only log the cname if it is non-empty, because `CNAME for service hello is ""` would be confusing.  It's fine also to add another log statement if there is no cname.

chbatey replied:
Done

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -2552,6 +2598,14 @@ func (s *AWSCloud) EnsureLoadBalancerDeleted(service *api.Service) error {
+		// Remove the DNS cname if defined
+		err := s.removeLoadBalancerCnameEntry(service, lb.DNSName)
+		if err != nil {
+			glog.Warningf("Error removing cname entry for %v: %v", service.Name, err)
```
I don't love that this is only a log warning, because it is easy to miss.  What would be the consequence of returning an error here?

chbatey replied:
K8s already spams the logs with a lot of errors that don't have any user impact. Where as this would leave a CName in R53 it doesn't stop any application working hence we went with a warning.

justinsb commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -308,3 +310,82 @@ func (s *AWSCloud) ensureLoadBalancerInstances(loadBalancerName string, lbInstan
+
+func getLoadBalancerHostedZone(service *api.Service) string {
+	if zone, ok := service.Annotations[ServiceAnnotationLoadBalancerCnameZone]; ok {
+		return zone + "."
```
I think this is going to fall foul of the "be liberal in what we accept" principle.  What if zone already had a dot on the end?  Also, it isn't clear why we choose to add a dot here - I suspect we should do it before talking to Route53 (which is why we want the dot, I believe)

justinsb commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -308,3 +310,82 @@ func (s *AWSCloud) ensureLoadBalancerInstances(loadBalancerName string, lbInstan
+		return ""
+	}
+
+	return fmt.Sprintf("%v-%v-%v.%v", service.Name, service.Namespace, s.cfg.Global.KubernetesClusterTag, hostedZone)
```
So is the idea here that we assign a stable name to the service, and that we then CNAME the actual hostname (`www.mycoolapp.com`) to that stable name?

chbatey replied:
Our main use case for this is to give a stable name to services for internal use.

We plan to use ingress for hostnames.

justinsb commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -308,3 +310,82 @@ func (s *AWSCloud) ensureLoadBalancerInstances(loadBalancerName string, lbInstan
+		return nil
+	}
+
+	s.r53Mutex.Lock()
```
Can you add a comment as to why this lock is needed?

justinsb commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -308,3 +310,82 @@ func (s *AWSCloud) ensureLoadBalancerInstances(loadBalancerName string, lbInstan
+	zonesInput := &route53.ListHostedZonesByNameInput{DNSName: aws.String(hostedZone)}
+	zones, err := s.r53.ListHostedZonesByName(zonesInput)
+	if err != nil {
+		return err
```
`return fmt.Errorf("error listing route53 zones: %v", err)` or something like that

Because this is the interface to third-party code, I like to wrap the errors.  Otherwise we end up with cryptic messages in the logs like "id not found" with no context.

justinsb commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -308,3 +310,82 @@ func (s *AWSCloud) ensureLoadBalancerInstances(loadBalancerName string, lbInstan
+	}
+
+	zone := zones.HostedZones[0]
+	if *zone.Name != hostedZone {
```
Safer to use `aws.StringValue(zone.Name) != hostedZone`, to avoid a panic just in case Name is nil

justinsb commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -308,3 +310,82 @@ func (s *AWSCloud) ensureLoadBalancerInstances(loadBalancerName string, lbInstan
+	}
+
+	if len(zones.HostedZones) == 0 {
+		return fmt.Errorf("No hosted zones found for account")
```
This could happen if hostedZone > existing zones, right?

Might be better to simply the logic into

`if len(zones.HostedZones) == 0 || aws.StringValue(zones.HostsZones[0].Name) != hostedZone`

justinsb commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -308,3 +310,82 @@ func (s *AWSCloud) ensureLoadBalancerInstances(loadBalancerName string, lbInstan
+	out, err := s.r53.ChangeResourceRecordSets(recordSetsInput)
+	if err != nil {
+		glog.Warningf("Failed to create dns entry: %v, %v", recordSetsInput.String(), out.String())
+		return err
```
`return fmt.Errorf(...)` as we are interfacing with non-k8s code.

justinsb commented:
Sorry it took me so long to get to this.

A few nits, but more importantly, is it sufficient to just add a single hostname derived from the service name.  What does it achieve for you vs using the ELB dns name itself?

I know this side-steps problems of establishing DNS ownership/control, but I'm just not entirely sure as to the goal here, if the user has to configure Route53 anyway.

ajohnstone commented:
@justinsb The ELB DNS name is not really comprehensible to people nor has a scheme that is easily determined being what looks like an arbitrary MD5 value (understandable considering a 32 char limit). Whilst waiting for this ticket I've side stepped the issue altogether by using aws lambda to generate the DNS.

Example ELB DNS:
```internal-ae033ca46070211e69b3a0a4b4f5db9b-761991085.eu-west-1.elb.amazonaws.com```

Example lambda:
```https://gist.github.com/ajohnstone/8f8c6ecdbbc6980597d0807331b779f2```

To me it makes more sense to be an ALIAS than a CNAME as you cannot set the APEX.
Secondly you have to manage an arbitrary DNS TTL with a CNAME.

One thing that would be a little odd is what happens if the same entry is used with a different namespace? And or another service?

My example use case is setting an &lt;service_name&gt;-&lt;namespace&gt;.&gt;hosted_zone&lt;
I've hyphenated the above to avoid issues with purchasing many SSL certs.

my-service-development.alias.photobox.com

This provides a scheme to which can be used by convention.

One issue is that the DNS creation is based on using cloudtrail, which is somewhat slow and can take a while to trigger the cloudwatch aws event to call the lambda. Anyway hopefully my use case will help a little. Just to note we don't expose kubernetes DNS internally.

On a final note, what happens to the DNS entry when the load balancer is terminated.

goltermann commented:
@chbatey can you rebase?  we're cutting the 1.3 branch for beta on Friday June 10

dchen1107 commented:
ping on rebase ...

@justinsb I removed 1.3 milestone from the pr since it hasn't been rebased & updated for a while. If it is required for aws, please add it back. Thanks!

chbatey commented:
Sorry for the delay. I have rebased and updated based on @justinsb's comments

david-mcmahon committed:
Kubernetes version v1.3.0-beta.0

david-mcmahon committed:
Versioning docs and examples for v1.3.0-beta.0.

david-mcmahon committed:
gofmt the file.

lavalamp committed:
Merge pull request #27408 from david-mcmahon/fix-basego

gofmt the file.

david-mcmahon committed:
versionize-docs.sh should call update-munge-docs.sh directly.

david-mcmahon committed:
Corrected build/versionize-docs.sh release-1.3.

lavalamp committed:
Merge pull request #27487 from david-mcmahon/fixall

Versioning docs and examples for release-1.3.

david-mcmahon committed:
Update api/swagger-spec files for the branch.

This is a new requirement this branch.

david-mcmahon committed:
Call hack/update-swagger-spec.sh too.

david-mcmahon committed:
Update api-reference html files too.

This was from a fresh run of the current build/versionize-docs.sh.
This should be tied to the latest corrected versionize-docs PR, but
we are fixing things one at a time here and sorting out the dependencies
as we go.

david-mcmahon committed:
update-generated-protobuf.sh for release-1.3.

lavalamp committed:
Merge pull request #27499 from david-mcmahon/fixswagger

Update api/swagger-spec files for the branch.

david-mcmahon committed:
Kubernetes version v1.3.0-beta.1

david-mcmahon committed:
Merge remote-tracking branch 'origin/master' into release-1.3

david-mcmahon committed:
build/versionize-docs.sh release-1.3

david-mcmahon committed:
Kubernetes version v1.3.0-beta.2

david-mcmahon committed:
Merge remote-tracking branch 'origin/master' into release-1.3

Commit:
Merge pull request #28030 from nikhiljindal/revertPR

Automatic merge from submit-queue

Revert "Federation e2e supports aws"

Reverting https://github.com/kubernetes/kubernetes/pull/27791 to get our Jenkins tests green again.

cc @kubernetes/sig-cluster-federation

(cherry picked from commit e93608c859db84b063814f95bd186593e40b354c)

Commit:
Merge pull request #28026 from quinton-hoole/2016-06-24-fix-nits-in-dns-deletion-logic-fix

Automatic merge from submit-queue

Address outstanding review comments in #27999.
(cherry picked from commit 9b215a33b53e9a12306c171cb28e7c0369e5e41e)

Commit:
Merge pull request #28034 from krousey/kubeconfig_filelock

Automatic merge from submit-queue

Adding lock files for kubeconfig updating

This is to prevent concurrent executions from corrupting the kubeconfig file.

Also, release note?

#23964

(cherry picked from commit cd422ada25ce25098145da3ccd1d3bf3d29bb47b)

Commit:
Merge pull request #28004 from mikedanese/fix-nil

Automatic merge from submit-queue

return nil from NewClientConfig instead of empty struct

This is a go convention and fixes an nil pointer in kubelet when passing in bad command line options:

```
I0624 04:12:33.333246   25404 plugins.go:141] Loaded network plugin "kubenet"
E0624 04:12:33.333390   25404 runtime.go:58] Recovered from panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:52
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:40
/usr/local/go/src/runtime/asm_amd64.s:472
/usr/local/go/src/runtime/panic.go:443
/usr/local/go/src/runtime/panic.go:62
/usr/local/go/src/runtime/sigpanic_unix.go:24
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/unversioned/service.go:132
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:254
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/client/cache/listwatch.go:80
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/client/cache/reflector.go:262
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/client/cache/reflector.go:204
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:86
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:87
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:49
```
cc @caesarxuchao @lavalamp
(cherry picked from commit 07d24e4557e4b3446c29fff48dd2d2dd685e627b)

Commit:
Merge pull request #28032 from fabioy/increase-reqs

Automatic merge from submit-queue

Increase pod CPU/memory for fluentd, dns and kube-proxy.

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

Improvements in cluster-wide resource utilization allows us to restore some of the CPU and memory limits on pods that was previously shrunk in 1.2.
(cherry picked from commit ef905c8125fa0b0b9c7657f5e04519f7470c10b8)

Commit:
Merge pull request #27208 from yujuhong/min_docker_version

Automatic merge from submit-queue

Bump minimum API version for docker to 1.21

The corresponding docker version is 1.9.x. Dropping support for docker 1.8.

/cc @kubernetes/sig-node
(cherry picked from commit 6f8a0c61c1fb057aab23d3f3010991fb35f28ac9)

Commit:
Merge pull request #28061 from mml/fed-double-quotes

Automatic merge from submit-queue

Remove extra double quotes in --federations.
(cherry picked from commit 440dfc6c207fae5a9103aaaadb8f8a70c05a4f9c)

Commit:
Merge pull request #28060 from yifan-gu/fix_stage1_anno

Automatic merge from submit-queue

rkt: Fix the 'privileged' check when stage1 annotation is provided.

Previously when stage1 annotation is provided, we only checks if
the kubelet allows privileged, which is not useful as that is a global
setting.

Instead, we should check if the pod has explicitly set the privileged
security context to 'true'.

cc @kubernetes/sig-rktnetes @kubernetes/sig-node
(cherry picked from commit f33188a0d0adcdc2896a31ccfdd4bd8bddf12fc4)

Commit:
Merge pull request #27996 from ronnielai/gc-threshold

Automatic merge from submit-queue

Image GC logic should compensate for reserved blocks

Calculating the disk usage based on available bytes instead of usage bytes to account for reserved blocks in image GC

#27169
(cherry picked from commit d50bd9e75c7a369dc51435605f919f602772af22)

Commit:
Merge pull request #28044 from yifan-gu/bump_rkt

Automatic merge from submit-queue

rkt: Bump required rkt version to 1.9.1.

Bump the rkt version to use the one that fixed the [go-systemd sdjournal issue](https://github.com/coreos/rkt/pull/2837), also match [release doc](https://github.com/kubernetes/kubernetes.github.io/blob/release-1.3/docs/getting-started-guides/rkt/index.md#prerequisite).

cc @kubernetes/sig-rktnetes @kubernetes/sig-node
(cherry picked from commit 151c2249ccd7b4b9a95b1656bb7354468855e45d)

Commit:
Merge pull request #28040 from ibm-contribs/fixaddons

Automatic merge from submit-queue

Tracked addition of federation, sed support in kube DNS

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

The kube DNS app recently gained support for federation (whatever that
is), including a new Salt parameter.  This broke the deployAddons.sh script for cluster ubuntu.  The DNS app also gained alternate
templates, intended to be friendly to `sed`.  Fortunately, those do
not demand a federation parameter.

This PR fixes up the ` cluster/ubuntu/deployAddons.sh` script to track those changes, by switching to the `sed`-friendly templates.
(cherry picked from commit 794dcc14f23cf9f148fd309b862bd0e7832f0561)

Commit:
Merge pull request #28043 from mml/fed-grace-period

Automatic merge from submit-queue

Set grace period to 0 when deleting namespaces after the test.

Otherwise, we try to run the next test and the pods are still there.
(cherry picked from commit 2bc2cf2e5edc34016d05243c4c49553c6024821d)

Commit:
Merge pull request #28002 from asalkeld/init-cache-error

Automatic merge from submit-queue

Fix startup type error in initializeCaches

The following error was getting logged:
PersistentVolumeController can't initialize caches, expected list of volumes, got:
&{TypeMeta:{Kind: APIVersion:} ListMeta:{SelfLink:/api/v1/persistentvolumes ResourceVersion:11} Items:[]}

The tests make extensive use of NewFakeControllerSource which uses api.List
instead of api.PersistentVolumeList. So use reflect to help iterate over the
items then assert the item type.

fixes #27757
(cherry picked from commit 1effc5af9119d16b9e3068dc89ac770e3bc441b2)

Commit:
Merge pull request #28087 from luxas/fix_hyperkube_dns

Automatic merge from submit-queue

Hotfix: Fixup the hyperkube dns manifest from a breaking federation PR

This one has to be cherrypicked into v1.3.0, ref: #27986

@girishkalele @madhusudancs @quinton-hoole @eparis @nikhiljindal @mml @roberthbailey @david-mcmahon
(cherry picked from commit d9cecdbc60711da5260247d3da31c6c0c6662432)

Commit:
Merge pull request #28108 from jsafrane/fix-cache-init

Automatic merge from submit-queue

Fix initialization of volume controller caches.

Fix `PersistentVolumeController.initializeCaches()` to pass pointers to volume or claim to  `storeObjectUpdate()` and add extra functions to enforce that the right types are checked in the future.

Fixes #28076
(cherry picked from commit be682148fb31628eefa8dc2da6f7bdf42f79b2ff)

Commit:
Merge pull request #27944 from wongma7/pvc-selector-hack

Automatic merge from submit-queue

Fix pvc label selector validation error

This is for https://github.com/kubernetes/kubernetes/issues/26866. Other selectors are e.g. v1beta1.x , the pvc one is unversioned.x
(cherry picked from commit 11e41581b42d4bfa771df19eebdf22f5032e81e8)

Commit:
Merge pull request #28056 from fabioy/increase-reqs

Automatic merge from submit-queue

Increase kube-dns requirements on CoreOS.

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

Missed changing the kube-dns memory limit on CoreOS. Follow of PR #28032.
(cherry picked from commit 812b87c8e607637532a4818d5a49368a3e501f0b)

Commit:
Merge pull request #28147 from jsafrane/clone-error-checks

Automatic merge from submit-queue

Fix error checks after cloning.

We should not use clone result When cloning fails. Fixes issues found in #28108.

(cherry picked from commit 8c5be726d81d091173697b84215e813f747a6557)

Commit:
Merge pull request #27513 from hpcloud/hpe/vsphere-vm-dir

Automatic merge from submit-queue

vSphere provider - Adding config for working dir

This allows the user the set "working-dir" in their vsphere.cfg file.
The value should be a path in the vSphere datastore in which the
provider will look for vms.  This should help compartmentalize
workloads in vSphere.
(cherry picked from commit bf54cd40f308bcd67de8e3407e17a1301780c5ec)

Commit:
Merge pull request #28159 from sttts/sttts-seccomp-operator

Automatic merge from submit-queue

Use : as seccomp security option operator for Docker 1.10

Compare https://docs.docker.com/v1.11/engine/reference/run/ and https://docs.docker.com/v1.10/engine/reference/run/

Fixes https://github.com/kubernetes/kubernetes/issues/28107
(cherry picked from commit 7b532f6f03d2462ed0dbba082359e378f46feb5d)

Commit:
Merge pull request #28165 from quinton-hoole/2016-06-22-fed-e2e-debugging

Automatic merge from submit-queue

Refactored, expanded and fixed federated-services e2e tests.

1. Moved BeforeEach() and AfterEach() to an inner scope, to prevent clashes with Framework's BeforeEach() and AfterEach().  Morte to come on this, as it's a major bug in our use of Ginkgo, and affects many other tests.
2. Keep track of which clusters we have created namespaces in, so that we don't try to delete namespaces out of clusters that we didn't create them in (e.g. the primary cluster, where the framework already creates and deleted the required namespace).
3. Separate tests for federated service creation and verification that underlying services are created correctly.
4. For DNS resolution tests, create backend pods (and delete on cleanup) where required).
5. For non-local DNS resolution, delete a backend pod in one cluster to test, and in the remainder of clusters on cleanup.
6. Lots of refactoring to make code re-usable across multiple test.
7. Lots of debugging/fixing to make sure that everything that the testscreate are cleaned up properly afterwards, and don't clash with the cleanups done by the e2e Framework.

(cherry picked from commit 1f628500074d7b3c680dc9faba1373387aef5d7c)

Commit:
Merge pull request #28095 from saad-ali/fixRacyVolumesInUse

Automatic merge from submit-queue

Kubelet should mark VolumeInUse before checking if it is Attached

Kubelet should mark VolumeInUse before checking if it is Attached.
Controller should fetch fresh copy of node object before detach instead of relying on node informer cache.

Fixes #27836

(cherry picked from commit 532491aab66467e1284fddf6e06c830b7f65da76)

j3ffml committed:
Merge pull request #28172 from zmerlynn/push-alt-ci

Build: Add KUBE_GCS_RELEASE_BUCKET_MIRROR option to push-ci-build.sh
(cherry picked from commit 43437e4528a48fb24d9515706f8a90e242d89e27)

zmerlynn committed:
Change references to gs://kubernetes-release/ci

Change over to gs://kubernetes-release-dev/ci. This should be all the
places we reference gs://kubernetes-release/ci or
https://storage.googleapis.com/kubernetes-release/ci. I'm happy to be
wrong.

Follow-on to #28172

Commit:
Merge pull request #28186 from nikhiljindal/betaAPI

Automatic merge from submit-queue

federation: Upgrading the groupversion to v1beta1

This PR contains 2 commits:
* Removing fields from Cluster API object that we are not using. This includes: Capacity, Allocatable and ClusterMeta.
* Move code and rename groupversion `federation/v1alpha1` to `federation/v1beta1`

cc @kubernetes/sig-cluster-federation
(cherry picked from commit 044de6ef901e9b05fa72ea710b83622c4465afa3)

mwielgus committed:
Merge pull request #28207 from mwielgus/ca-0.2.2

Bump cluster autoscaler to 0.2.2
(cherry picked from commit d7eaad95c47d0c1a358ae867aa4bd975b224ad97)

Commit:
Merge pull request #28160 from saad-ali/waitForGracefulTerm

Automatic merge from submit-queue

Volume manager must verify containers terminated before deleting for ungracefully terminated pods

A pod is removed from volume manager (triggering unmount) when it is deleted from the kubelet pod manager. Kubelet deletes the pod from pod manager as soon as it receives a delete pod request. As long as the graceful termination period is non-zero, this happens after kubelet has terminated all containers for the pod. However, when graceful termination period for a pod is set to zero, the volume is deleted from pod manager *before* its containers are terminated.

This  can result in volumes getting unmounted from a pod before all containers have exited when graceful termination is set to zero.

This PR prevents that from happening by only deleting a volume from volume manager once it is deleted from the pod manager AND the kubelet containerRuntime status indicates all containers for the pod have exited. Because we do not want to call containerRuntime too frequently, we introduce a delay in the `findAndRemoveDeletedPods()` method to prevent it from executing more frequently than every two seconds.

Fixes https://github.com/kubernetes/kubernetes/issues/27691

Running test in tight loop to verify fix.

(cherry picked from commit c6fa861688e7ad4dcf5ced968a2e525ce5c7969f)

Commit:
Merge pull request #28211 from quinton-hoole/2016-06-29-fix-federation-e2e

Automatic merge from submit-queue

Fix federation e2e tests by correctly managing cluster clients

1. The main fix: Correct overall BeforeEach() to create a new set of cluster clients, rather than just append to the set created by all previous tests.  This was screwing up a lot of stuff in difficult to diagnose ways.
2. Add lots of debug logging.
3. Be better about cleaning up after each test.

```
SUCCESS! -- 6 Passed | 0 Failed :-)
```

cc @nikhiljindal @madhusudancs @mfanjie @colhom FYI
(cherry picked from commit 02de0092a012b6b5cd434eb7863ab62374248c3d)

cjcullen committed:
Lock all possible kubecfg files at the beginning of ModifyConfig.

Commit:
Merge pull request #28132 from madhusudancs/fed-kubedns-flags-nodebootstrap

Automatic merge from submit-queue

Substitute federation_domain_map parameter with its value in node bootstrap scripts.

This PR also removes the substitution code we added to the build scripts.

**Release Note**

```release-note
If you use one of the kube-dns replication controller manifest in `cluster/saltbase/salt/kube-dns`, i.e. `cluster/saltbase/salt/kube-dns/{skydns-rc.yaml.base,skydns-rc.yaml.in}`, either substitute one of `__PILLAR__FEDERATIONS__DOMAIN__MAP__` or `{{ pillar['federations_domain_map'] }}` with the corresponding federation name to domain name value or remove them if you do not support cluster federation at this time. If you plan to substitute the parameter with its value, here is an example for `{{ pillar['federations_domain_map'] }`
pillar['federations_domain_map'] = "- --federations=myfederation=federation.test"
where `myfederation` is the name of the federation and `federation.test` is the domain name registered for the federation.
```

cc @erictune  @kubernetes/sig-cluster-federation @MikeSpreitzer @luxas

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

(cherry picked from commit 3a6494e9aeab8878f38f3367708fe80d7e6cf773)

Commit:
Merge pull request #28119 from eparis/release-1.3

Automatic merge from submit-queue

Batch update for 1.3

#28030: Revert "Federation e2e supports aws"
#28026: Address outstanding review comments in #27999.
#28034: Adding lock files for kubeconfig updating
#28004: return nil from NewClientConfig instead of empty struct
#28032: Increase pod CPU/memory for fluentd, dns and kube-proxy.
#27208: Bump minimum API version for docker to 1.21
#28061: Remove extra double quotes in --federations.
#28060: rkt: Fix the 'privileged' check when stage1 annotation is provided.
#27996: Image GC logic should compensate for reserved blocks
#28044: rkt: Bump required rkt version to 1.9.1.
#28040: Tracked addition of federation, sed support in kube DNS
#28043: Set grace period to 0 when deleting namespaces after the test.
#28002: Fix startup type error in initializeCaches
#28087: Hotfix: Fixup the hyperkube dns manifest from a breaking federation PR
#28108: Fix initialization of volume controller caches.
#28056: Increase kube-dns requirements on CoreOS.
#28147: Fix error checks after cloning.
#28159: Use : as seccomp security option operator for Docker 1.10
#28165: Refactored, expanded and fixed federated-services e2e tests.
#28095: Kubelet should mark VolumeInUse before checking if it is Attached
#28172: Build: Add KUBE_GCS_RELEASE_BUCKET_MIRROR option to push-ci-build.sh
#28207: Bump cluster autoscaler to 0.2.2
#28160: Volume manager must verify containers terminated before deleting for ungracefully terminated pods
#28211: Fix federation e2e tests by correctly managing cluster clients
#27944: Fix pvc label selector validation error
#28186: federation: Upgrading the groupversion to v1beta1

Commit:
Merge pull request #28261 from mml/fed-etcd-pv

Automatic merge from submit-queue

Create a PD volume when deploying the federation with federation-up.
(cherry picked from commit 1d9b61493d2668886a40f1550d1c9fd592ec911b)

Commit:
Merge pull request #25873 from simonswine/fix-flocker-restart-master

Automatic merge from submit-queue

Fix problems with container restarts and flocker volumes

[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/.github/PULL_REQUEST_TEMPLATE.md?pixel)]()

The issue is described in #22436. A pod with a flocker volume attached can't be restarted. A complete new pod has to be created, then everything is working fine.

The problem is around the `isReady` check. This ready state is not getting cleaned up for restarting containers. IMHO we don't need this ready state stored at all. Maybe it was meant to not call the flocker API during pod restarts, but I would rather have it removed... (happy about some hints why we need this lines)

I tested it in a AWS/flocker cluster: starting/restarting/rescheduling on another node (in the same zone) works perfectly fine.

(cherry picked from commit b1b51a6fefada3e2a4167c1fd7bc986b6608b267)

Commit:
Merge pull request #28258 from vishh/28231

Automatic merge from submit-queue

[kubelet] Allow opting out of automatic cloud provider detection in kubelet. By default kubelet will auto-detect cloud providers

fixes #28231
(cherry picked from commit ab37fbf4c268b6b0eca4d3aba0284480a972c754)

erictune committed:
Merge pull request #28262 from eparis/release-1.3

Release 1.3 - batch 2

david-mcmahon committed:
Kubernetes version v1.3.0-beta.3

david-mcmahon committed:
Kubernetes version v1.3.0

david-mcmahon committed:
Kubernetes version v1.3.1-beta.0

erictune committed:
Add some whitespace to test CI/CD

erictune committed:
Merge pull request #28386 from erictune/luvwhitespace

Add some whitespace to test CI/CD

gmarek committed:
Move CIDR allocation logic away from nodecontroller.go

zmerlynn committed:
Migrate to kuberetes-release-dev

This turns off the kubernetes-release-dev mirror. #28193 should have
shifted all the references over, this should ensure that it's true.

zmerlynn committed:
Merge pull request #28783 from zmerlynn/automated-cherry-pick-of-#28193-upstream-release-1.3

Automated cherry pick of #28193

Commit:
Merge pull request #28841 from zmerlynn/automated-cherry-pick-of-#28792-upstream-release-1.3

Automatic merge from submit-queue

Automated cherry pick of #28792

Cherry pick of #28792 on release-1.3.

Commit:
Merge pull request #28397 from cjcullen/automated-cherry-pick-of-#28232-upstream-release-1.3

Automatic merge from submit-queue

Automated cherry pick of #28232

Cherry pick of #28232 on release-1.3.

liggitt committed:
Fix watch cache filtering

Commit:
Merge pull request #28968 from liggitt/cache-filter-release-1.3

Automatic merge from submit-queue

release-1.3: Fix watch cache filtering

Cherry-pick of https://github.com/kubernetes/kubernetes/pull/28966

david-mcmahon committed:
Kubernetes version v1.3.2-beta.0

bprashanth committed:
List all nodes and occupy cidr map before starting allocations

bprashanth committed:
Fix integration test for nodecontroller cherrypick

Commit:
Merge pull request #29063 from bprashanth/automated-cherry-pick-of-#28604-#29062-upstream-release-1.3

Automatic merge from submit-queue

Automated cherry pick of #28604 #29062

Cherry pick of #28604 #29062 on release-1.3.

david-mcmahon committed:
Kubernetes version v1.3.2

MrHohn commented on pkg/cloudprovider/providers/aws/aws_loadbalancer.go:
```diff
@@ -443,3 +445,82 @@ func proxyProtocolEnabled(backend *elb.BackendServerDescription) bool {
+	return s.updateLoadBalancerCnameEntry(service, dnsName, "UPSERT")
+}
+
+func (s *AWSCloud) removeLoadBalancerCnameEntry(service *api.Service, dnsName *string) error {
```
Why removeLoadBalancerCnameEntry() is never called except in aws_test.go?
